### PR TITLE
Decouple NamedXContentRegistry from ElasticsearchException

### PR DESCRIPTION
**Pull request by dakrone**
This commit decouples `NamedXContentRegistry` from using either
`ElasticsearchException`, `ParsingException`, or `UnknownNamedObjectException`.

This will allow us to move NamedXContentRegistry to its own lib as part of the
xcontent extraction work.

Relates to #28504

### CODE REVIEW

**Review by cbuescher — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java [hunk 1]
```diff
@@ -316,11 +317,11 @@ public static QueryBuilder parseInnerQueryBuilder(XContentParser parser) throws
         QueryBuilder result;
         try {
             result = parser.namedObject(QueryBuilder.class, queryName, null);
-        } catch (UnknownNamedObjectException e) {
+        } catch (XContentParseException e) {
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: If we're going to use this `XContentParseException` for anything other that unknown named objects we shouldn't catch it here. We need to be sure this is an unknown named object. Either we should rename the exception or make a subclass of it for unknown named objects, I think.

Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java]: Yeah that's probably a good idea, for now there are no other uses of this exception, but if that changes in the future I will make a subclass of it for this.

File: server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java [hunk 2]
```diff
@@ -1007,22 +1008,20 @@ public void testNamedObject() throws IOException {
         {
             p.nextToken();
             assertEquals("test", p.namedObject(Object.class, "str", null));
-            UnknownNamedObjectException e = expectThrows(UnknownNamedObjectException.class,
+            XContentParseException e = expectThrows(XContentParseException.class,
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: I think this should be `NamedObejectNotFoundException` instead, right? It is more specific.

File: server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java [hunk 3]
```diff
@@ -1007,22 +1008,20 @@ public void testNamedObject() throws IOException {
...
                     () -> p.namedObject(Object.class, "unknown", null));
-            assertEquals("Unknown Object [unknown]", e.getMessage());
-            assertEquals("java.lang.Object", e.getCategoryClass());
-            assertEquals("unknown", e.getName());
+            assertThat(e.getMessage(), endsWith("unable to parse Object with name [unknown]: parser not found"));
         }
         {
-            Exception e = expectThrows(ElasticsearchException.class, () -> p.namedObject(String.class, "doesn't matter", null));
-            assertEquals("Unknown namedObject category [java.lang.String]", e.getMessage());
+            Exception e = expectThrows(XContentParseException.class, () -> p.namedObject(String.class, "doesn't matter", null));
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: Here too

File: server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java [hunk 4]
```diff
@@ -1007,22 +1008,20 @@ public void testNamedObject() throws IOException {
...
+            assertEquals("unknown named object category [java.lang.String]", e.getMessage());
         }
         {
             XContentParser emptyRegistryParser = xcontentType().xContent()
                 .createParser(NamedXContentRegistry.EMPTY, DeprecationHandler.THROW_UNSUPPORTED_OPERATION, new byte[] {});
-            Exception e = expectThrows(ElasticsearchException.class,
+            Exception e = expectThrows(XContentParseException.class,
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/common/xcontent/BaseXContentTestCase.java]: And here!

File: server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java [hunk 5]
```diff
@@ -220,8 +221,8 @@ public void testUnknownFieldsExpection() throws IOException {
                 "}\n";
         {
             XContentParser parser = createParser(rescoreElement);
-            Exception e = expectThrows(ParsingException.class, () -> RescorerBuilder.parseFromXContent(parser));
-            assertEquals("Unknown RescorerBuilder [bad_rescorer_name]", e.getMessage());
+            Exception e = expectThrows(XContentParseException.class, () -> RescorerBuilder.parseFromXContent(parser));
```
Comment (nik9000) [inline server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java]: All of these should be `NamedObjectNotFoundException` I think.

Comment (dakrone) [reply server/src/test/java/org/elasticsearch/search/rescore/QueryRescorerBuilderTests.java]: Yeah, I changed all the occurences, thanks

**Review by nik9000 — APPROVED** [review]
:+1:

**Pull request merged**